### DISCUSSION TITLE
I might be interested

### DISCUSSION THREAD

**Discussion by plgibbons-lenovo** [microsoft/mu_feature_uefi_variable#98] [General]
Hi, mu variable team. I just ran across this and immediately got sucked in after reading the design doc. I would be interested in exploring more about what is going on in this project to see if I might want to contribute (if I can make the time to do so). 
I don't see any posts yet this year. What is the current status?
Are you having regular design meetings?

**Comment by makubacki** [comment]
Thank you for your interest. We put this work on hold. We're not sure when it will start back up again, but we'll definitely reach out to you when it does.

Is there anything in particular you were interested in?

**Reply by plgibbons-lenovo** [reply]
There isn't a particular part that I would like to work on, if that is what you are asking. I would like to see this come together so we might be able to use it in our product someday, plus the description of the architecture seems intriguing. A lot of the code in EDK2 kinda scares me - was mention here, tech debt, like overly large functions with lack of unit testing, etc. I was a little surprised by the meeting note that said you don't intend to maintain public interface compatibility with the current UEFI spec. Did I understand that correctly?

**Reply by makubacki** [reply]
That's likely. The current UEFI variable expectations are written around the assumption of properties provided by SPI flash. Today, UEFI variables are being stored on eMMC, UFS, and off-chip engines. In those cases, some requirements like storing data to non-volatile storage before returning from a `SetVariable()` call might not be practical. Example, early in the host OS load process when a NAND driver stack is not available, but the kernel wants to write a variable.

Also, we'd like to bring more focus to the security properties of UEFI variables. Instead of choosing when to lock them with a separate API (that is not required) we'd like security policy to be part of the variable creation call and mandatory to set a variable.

The solution breaks down `user-facing APIs` and elements in the control flow like `gaskets`. It is possible, we could have an API that mirrors the existing API and use a gasket to flow the data into a stack that's also leveraged by the new API.

In general though, if we rewrite the stack, we don't want to be limited by the pre-existing API.

**Reply by plgibbons-lenovo** [reply]
I was shocked when I was first learning about EFI and there seemed to have been no thought given to securing 'EFI environment variables' in those days. As for your example of adherence to the current spec, is the thought to let the interface advertise semantics and let the caller decide if they want to use it? For example (your example), the semantics might allow/require writes to be posted for later commitment to NV storage with some kind of status that can be gather later. That might not suit all callers. Not sure what they would do instead though. It wasn't clear to me how you intended to support multiple back-ends simultaneously (assuming you are). Again, the interface could advertise, or callers could request certain semantics for guarantees for speed/security/tamper evidence/redundancy - whatever properties/profiles this subsystem might want to define/allow and back-ends might want to support.
I hope I am not spamming you here.

**Reply by makubacki** [reply]
The user-facing API is still TBD and thought needs to be given to clearly defining expectations with callers.

There are various ways to approach this as you highlighted by loosening interface guarantees, for example, or allowing backends to influence the capabilities reported to callers. This is a reason the user-facing APIs have not been formally proposed yet.

As another example, kernels may attempt to store breadcrumbs in UEFI variables during panic scenarios because they might have either lost or do not trust their storage driver stack. That might simply be a use case that is not supported depending on a particular backend. Today, that is brought to caller's attention most often by an error status code being returned to the caller. The variable implementation cannot change availability of backends, but I would like to make it clear at the interface level what is expected to reasonably work or not work so calling software can factor that into its design.

Some of these things break down into common categories, like today only `synchronous` variables are supported. Alternatively, we've talked about `asynchronous` flows. Regardless of the particular backend, that distinction could be made apparent in the interface and expectations set accordingly.

Simultaneous support of backends would be accomplished via a `router` built into the variable stack that has a `policy` assigned to it by the platform. The policy dictates how to route variables based on properties like name, vendor GUID, attributes, etc. A policy could be as simple as route all variables to SPI flash or route variables with this vendor GUID to this backend and everything else to SPI flash, for example. Like the variable driver today, a memory cache would be maintained that represents a backend. If the backend write succeeds, the memory cache is updated.
